rfd/0199-hardware-key-agent.md

```diff
@@ -0,0 +1,685 @@
+prompts for the user as needed without adding additional overhead.
+
+Alternative: By default, set `teleport.agent=false`. Once the user logs into
+Connect with a hardware key requirement for the first time, flip the flag to
+true indefinitely. The benefit with this approach is that we won't run an
+unused agent by default for users not using hardware key support. On the other
+hand, it requires the user to log in at least once for the feature to work as
+described in the example above.
+
+If desired, the agent can be disabled manually with a config option:
+
+| Property | Default | Description |
+|----------|---------|-------------|
+| `hardwareKeyAgent.enabled` | `false` | Starts the hardware key agent automatically |
+
+##### `tsh agent`
+
+`tsh agent` will be made available as a hidden command, primarily for
+development. If in the future we get requests to fully support this command,
+we may make it a public command and make any necessary improvements.
+
+For now, we will not put exorbitant effort into providing a good UX with this command
+(e.g. stealing focus for hardware key prompts, re-login on cert expiration),
+but it should be fully functional.
+
+##### Agent already running
+
+If there is already a hardware key agent running at `$TEMP/.Teleport-PIV/agent.sock`,
+a newly started agent client would fail to open a new listener on that same
+socket. Instead of failing, the new client will ping the running agent to check
+if it's active.
+
+If the ping request fails, the socket will be treated as abandoned and
+automatically replaced by the new agent.
+
+If the ping request succeeds, an error will be returned:
+
+```console
+> tsh piv agent
+Error: another agent instance is already running. PID: 86946.
+```
+
+Note: this should be a very uncommon edge case, as it would only occur if the
+user already has `tsh agent` running or another instance of Teleport connect.
+
+Note: with Teleport Connect, this error would be displayed when it attempts
+to start the agent, but Teleport Connect would not fail to start. The error
+would be shown in Teleport Connect's debug logs.
+
+##### Running the agent before login
+
+The hardware key agent does not depend on Teleport login state, meaning a user
+can run it before logging in.
+
+For the login itself, Teleport clients will interface directly with the
+hardware key to check/generate the private key. Then, the client will interface
+with the key through the hardware key agent that is already running.
+
+#### Dependent clients
+
+Dependent clients should interact with the hardware key agent seamlessly so
+that there is limited UX degradation when compared to the a client connecting
+directly to the hardware key.
+
+##### Agent stops running
+
+If the agent stops unexpectedly during a dependent client's operation, it may
+lead to an error for the dependent client. Rather than returning this error,
+the client will log the error as debug and try again by connecting directly to
+the hardware key.
```

Comment:
How do we get out of this failure mode and get a new agent client running again? We don't want one error to cause all clients to start hitting the hardware key directly and rendering the new agent functionality useless.

Reply:
It's not so much an error, as the agent is no longer running. e.g. the user closes Teleport Connect, or cancels `tsh piv agent`. My point here is just that if a user runs a `tsh` command as the agent is being closed, we should make a best effort attempt at completing the command instead of returning an error like `could not connect to hardware key agent`.

Reply:
I've added a note that the agent will attempt to restart with backoff, but in the interim clients would still fallback go direct connections. I was originally thinking of a situation where the agent process was force killed and left an unserviced socket, etc.